MediaBrowser.Common/Net/IPHost.cs

```diff
@@ -0,0 +1,444 @@
+                }
+            }
+            else
+            {
+                output = HostName;
+            }
+
+            return output;
+        }
+
+        /// <inheritdoc/>
+        public override void Remove(AddressFamily family)
+        {
+            if (ResolveHost())
+            {
+                _addresses = _addresses.Where(p => p.AddressFamily != family).ToArray();
+            }
+        }
+
+        /// <inheritdoc/>
+        public override bool Contains(IPObject address)
+        {
+            // An IPHost cannot contain another IPObject, it can only be equal.
+            return Equals(address);
+        }
+
+        /// <inheritdoc/>
+        protected override IPObject CalculateNetworkAddress()
+        {
+            var netAddr = NetworkAddressOf(this[0], PrefixLength);
+            return new IPNetAddress(netAddr.Address, netAddr.PrefixLength);
+        }
+
+        /// <summary>
+        /// Attempt to resolve the ip address of a host.
+        /// </summary>
+        /// <returns><c>true</c> if any addresses have been resolved, otherwise <c>false</c>.</returns>
+        private bool ResolveHost()
+        {
+            // When was the last time we resolved?
+            if (_lastResolved == 0)
+            {
+                _lastResolved = DateTime.UtcNow.Ticks;
+            }
+
+            // If we haven't resolved before, or out timer has run out...
+            if ((_addresses.Length == 0 && !Resolved) || (TimeSpan.FromTicks(DateTime.UtcNow.Ticks - _lastResolved).TotalMinutes > Timeout))
+            {
+                _lastResolved = DateTime.UtcNow.Ticks;
+                ResolveHostInternal().GetAwaiter().GetResult();
+                Resolved = true;
+            }
+
+            return _addresses.Length > 0;
+        }
+
+        /// <summary>
+        /// Task that looks up a Host name and returns its IP addresses.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private async Task ResolveHostInternal()
+        {
+            if (!string.IsNullOrEmpty(HostName))
+            {
+                // Resolves the host name - so save a DNS lookup.
+                if (string.Equals(HostName, "localhost", StringComparison.OrdinalIgnoreCase))
+                {
+                    _addresses = new IPAddress[] { new IPAddress(Ipv4Loopback), new IPAddress(Ipv6Loopback) };
+                    return;
+                }
+
+                if (Uri.CheckHostName(HostName).Equals(UriHostNameType.Dns))
+                {
+                    try
+                    {
+                        IPHostEntry ip = await Dns.GetHostEntryAsync(HostName).ConfigureAwait(false);
+                        _addresses = ip.AddressList;
+                    }
+                    catch (SocketException)
+                    {
+                        // Ignore socket errors, as the result value will just be an empty array.
```

Comment:
But do we need to log?

Reply:
In an ideal world it might be nice - however, there isn't a logger object accessible at this point in the code, and the code expects that dns won't be available all of the time.

Will give it some thought to how this is possible.

Reply:
`Trace` or `Debug`?